File: backend/app/api/meal_analysis.py
from __future__ import annotations

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user, get_meal_analysis_service
from app.models.user import User
from app.repositories.nutrition import NutritionRepository
from app.schemas.meal_analysis import (
    AnalysisConfirm,
    AnalysisResponse,
    TextAnalysisRequest,
    VoiceAnalysisResponse,
)
from app.schemas.nutrition import MealItemResponse
from app.services.meal_analysis import MealAnalysisService
from app.services.nutrition import NutritionService

router = APIRouter(prefix="/nutrition/analyze", tags=["Nutrition AI"])


@router.post(
    "/text",
    response_model=AnalysisResponse,
    summary="Estimate macros from a free-form description",
)
async def analyze_text(
    data: TextAnalysisRequest,
    service: MealAnalysisService = Depends(get_meal_analysis_service),
    _: User = Depends(get_current_user),
) -> AnalysisResponse:
    return await service.analyze_text(data.text)


@router.post(
    "/photo",
    response_model=AnalysisResponse,
    summary="Estimate macros from a meal photo",
)
async def analyze_photo(
    file: UploadFile = File(...),
    note: str | None = Form(None),
    service: MealAnalysisService = Depends(get_meal_analysis_service),
    _: User = Depends(get_current_user),
) -> AnalysisResponse:
    payload = await file.read()
    if not payload:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Empty file upload",
        )
    return await service.analyze_photo(
        payload, file.content_type or "image/jpeg", hint=note or None
    )


@router.post(
    "/voice",
    response_model=VoiceAnalysisResponse,
    summary="Transcribe an audio note and estimate macros from the transcript",
)
async def analyze_voice(
    file: UploadFile = File(...),
    service: MealAnalysisService = Depends(get_meal_analysis_service),
    _: User = Depends(get_current_user),
) -> VoiceAnalysisResponse:
    payload = await file.read()
    if not payload:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Empty file upload",
        )
    return await service.analyze_voice(
        payload, file.content_type or "audio/webm"
    )


@router.post(
    "/{meal_id}/confirm",
    response_model=MealItemResponse,
    status_code=201,
    summary="Persist a confirmed AI prediction inside an existing meal",
)
async def confirm_analysis(
    meal_id: int,
    data: AnalysisConfirm,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> MealItemResponse:
    nutrition = NutritionService(NutritionRepository(db))
    return await nutrition.add_item_from_analysis(
        meal_id, current_user.id, data
    )

File: frontend/src/api/mealAnalysis.ts
import { AnalysisConfirm, AnalysisResponse, MealItem, VoiceAnalysisResponse } from "../types";
import api from "./client";

export const analyzeText = (text: string) =>
  api.post<AnalysisResponse>("/nutrition/analyze/text", { text });

export const analyzePhoto = (file: File, note?: string) => {
  const fd = new FormData();
  fd.append("file", file);
  if (note && note.trim()) fd.append("note", note.trim());
  // Intentionally do NOT set Content-Type — axios appends the multipart
  // boundary when the body is a FormData instance.
  return api.post<AnalysisResponse>("/nutrition/analyze/photo", fd);
};

export const analyzeVoice = (file: Blob, filename = "voice.webm") => {
  const fd = new FormData();
  fd.append("file", file, filename);
  return api.post<VoiceAnalysisResponse>("/nutrition/analyze/voice", fd);
};

export const confirmAnalysis = (mealId: number, payload: AnalysisConfirm) =>
  api.post<MealItem>(`/nutrition/analyze/${mealId}/confirm`, payload);
